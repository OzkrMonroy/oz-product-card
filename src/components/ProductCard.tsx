import React, { createContext } from 'react';
import styles from '../styles/styles.module.css';
import {
  ProductCardProps,
  ProductContextProps,
} from '../interfaces/productInterfaces';
import { useProduct } from '../hooks/useProduct';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export const ProductCard = ({
  product,
  children,
  className,
  style,
  onChange,
  count,
  initialValues,
}: ProductCardProps) => {
  const { counter, increaseBy, isMaxCountReached, reset } = useProduct({
    product,
    onChange,
    initialCounter: count,
    initialValues,
  });

  return (
    <Provider
      value={{
        product,
        counter,
        increaseBy,
        maxCount: initialValues?.maxCount,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          reset,
          maxCount: initialValues?.maxCount,
          product,
          increaseBy,
        })}
        {/* <ProductImage img={img} name={name} />
      <ProductTitle name={name} />
      <ProductsButtons counter={counter} increaseBy={increaseBy}/> */}
      </div>
    </Provider>
  );
};
