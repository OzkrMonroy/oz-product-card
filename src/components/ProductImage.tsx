import React, { useContext } from 'react';
import { ProductContext } from './ProductCard';
import { ProductImageProps } from '../interfaces/productInterfaces';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

export const ProductImage = ({
  img,
  name,
  className,
  style,
}: ProductImageProps) => {
  const { product } = useContext(ProductContext);
  let imageToShow: string;
  let titleForAlt: string;

  if (img) {
    imageToShow = img;
  } else if (product.img) {
    imageToShow = product.img;
  } else {
    imageToShow = noImage;
  }

  if (name) {
    titleForAlt = name;
  } else if (product.name) {
    titleForAlt = product.name;
  } else {
    titleForAlt = 'Product';
  }

  return (
    <img
      src={imageToShow}
      alt={titleForAlt}
      className={`${styles.productImg} ${className}`}
      style={style}
    />
  );
};
