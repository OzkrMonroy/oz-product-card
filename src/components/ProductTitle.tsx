import React, { useContext } from 'react';
import { ProductContext } from './ProductCard';
import { ProductTitleProps } from '../interfaces/productInterfaces';
import styles from '../styles/styles.module.css';

export const ProductTitle = ({ name, className, style }: ProductTitleProps) => {
  const { product } = useContext(ProductContext);
  return (
    <span className={`${styles.productDescription} ${className}`} style={style}>
      {name ? name : product.name}
    </span>
  );
};
