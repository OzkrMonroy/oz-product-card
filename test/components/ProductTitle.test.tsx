import React from 'react';
import renderer from 'react-test-renderer';
import { ProductTitle } from '../../src/components';
import ProductCard from '../../src/components/index';
import { product } from '../mocks/products';

describe('Product Title component', () => {
  test('should render the title which is provide by prop', () => {
    const wrapper = renderer.create(<ProductTitle name="Custom name" />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
  test('should display the name of the product that is saved in the context', () => {
    const wrapper = renderer.create(
      <ProductCard product={product}>{() => <ProductTitle />}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
