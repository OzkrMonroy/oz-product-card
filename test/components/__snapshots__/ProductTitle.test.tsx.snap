// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Title component should display the name of the product that is saved in the context 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    This is an example - Without image
  </span>
</div>
`;

exports[`Product Title component should render the title which is provide by prop 1`] = `
<span
  className="productDescription undefined"
>
  Custom name
</span>
`;
